Haskell extraction is broken for ghc 9.0

```coq
Require Import Coq.extraction.ExtrHaskellBasic.
Extraction Language Haskell.
Definition foo (t : { T : Type & T }) := projT2 t.
Definition bar := foo (existT _ (nat:Type) 0).
Recursive Extraction bar.
(*
{-# OPTIONS_GHC -cpp -XMagicHash #-}
{- For Hugs, use the option -F"cpp -P -traditional" -}

module Main where

import qualified Prelude

#ifdef __GLASGOW_HASKELL__
import qualified GHC.Base
import qualified GHC.Prim
#else
-- HUGS
import qualified IOExts
#endif

#ifdef __GLASGOW_HASKELL__
unsafeCoerce :: a -> b
unsafeCoerce = GHC.Base.unsafeCoerce#
#else
-- HUGS
unsafeCoerce :: a -> b
unsafeCoerce = IOExts.unsafeCoerce
#endif

#ifdef __GLASGOW_HASKELL__
type Any = GHC.Prim.Any
#else
-- HUGS
type Any = ()
#endif

__ :: any
__ = Prelude.error "Logical or arity value used"

data Nat =
   O
 | S Nat

data SigT a p =
   ExistT a p

projT2 :: (SigT a1 a2) -> a2
projT2 x =
  case x of {
   ExistT _ h -> h}

foo :: (SigT () Any) -> Any
foo t =
  projT2 t

bar :: Nat
bar =
  unsafeCoerce foo (ExistT __ (unsafeCoerce O))

*)
```
This code does not compile with ghc 9.0.1, failing with 
```
Error:     Not in scope: ‘GHC.Base.unsafeCoerce#’
    Perhaps you meant ‘GHC.Base.unsafeChr’ (imported from GHC.Base)
    Module ‘GHC.Base’ does not export ‘unsafeCoerce#’.
   |
22 | unsafeCoerce = GHC.Base.unsafeCoerce#
   |                ^^^^^^^^^^^^^^^^^^^^^^
```

According to https://github.com/haskell/primitive/issues/258, we should be using `GHC.Exts.unsafeCoerce#`, but the diff at https://gitlab.haskell.org/ghc/ghc/commit/74ad75e87317196c600dfabc61aee1b87d95c214 suggests this is not backwards compatible.

Also, it seems extraction is already broken on earlier versions of Haskell, as I get
```
/tmp/foo.hs:26:12: error:
    Not in scope: type constructor or class ‘GHC.Prim.Any’
    Perhaps you meant ‘GHC.Prim.Int#’ (imported from GHC.Prim)
    Module ‘GHC.Prim’ does not export ‘Any’.
   |
26 | type Any = GHC.Prim.Any
   |            ^^^^^^^^^^^^
```
with 8.6.5.

cc, uh, @zeldovich ?  (I'm not sure who knows about Hakell extraction)

> ```
> /tmp/foo.hs:26:12: error:
>     Not in scope: type constructor or class ‘GHC.Prim.Any’
>     Perhaps you meant ‘GHC.Prim.Int#’ (imported from GHC.Prim)
>     Module ‘GHC.Prim’ does not export ‘Any’.
>    |
> 26 | type Any = GHC.Prim.Any
>    |            ^^^^^^^^^^^^
>  ```

The `GHC.Prim.Any` was already addressed in #6023 from Coq 8.7 (2017). The patch could easily be backported to the dev branch of Coq 8.6 if needed but I'm afraid there would be no further release of the 8.6 branch...

I made #14346 to fix unsafeCoerce (not tested).

--- a/doc/changelog/10-standard-library/14345-new-haskell.rst
+++ b/doc/changelog/10-standard-library/14345-new-haskell.rst
@@ -1,0 +1,5 @@
+- **Fixed:**
+  Haskell extraction is now compatible with GHC versions >= 9.0.  Some ``#if``
+  statements have been added to extract ``unsafeCoerce`` to its new location in
+  newer versions of GHC.  (`#14345 <https://github.com/coq/coq/pull/14345>`_,
+  fixes `#14256 <https://github.com/coq/coq/issues/14256>`_, by Jason Gross).

--- a/plugins/extraction/haskell.ml
+++ b/plugins/extraction/haskell.ml
@@ -57,6 +57,9 @@ let preamble mod_name comment used_modules usf =
    else
      str "#ifdef __GLASGOW_HASKELL__" ++ fnl () ++
      str "import qualified GHC.Base" ++ fnl () ++
+     str "#if __GLASGOW_HASKELL__ >= 900" ++ fnl () ++
+     str "import qualified GHC.Exts" ++ fnl () ++
+     str "#endif" ++ fnl () ++
      str "#else" ++ fnl () ++
      str "-- HUGS" ++ fnl () ++
      str "import qualified IOExts" ++ fnl () ++
@@ -66,7 +69,11 @@ let preamble mod_name comment used_modules usf =
    else
      str "#ifdef __GLASGOW_HASKELL__" ++ fnl () ++
      str "unsafeCoerce :: a -> b" ++ fnl () ++
+     str "#if __GLASGOW_HASKELL__ >= 900" ++ fnl () ++
+     str "unsafeCoerce = GHC.Exts.unsafeCoerce#" ++ fnl () ++
+     str "#else" ++ fnl () ++
      str "unsafeCoerce = GHC.Base.unsafeCoerce#" ++ fnl () ++
+     str "#endif" ++ fnl () ++
      str "#else" ++ fnl () ++
      str "-- HUGS" ++ fnl () ++
      str "unsafeCoerce :: a -> b" ++ fnl () ++

--- a/test-suite/output/Extraction_Haskell_String_12258.out
+++ b/test-suite/output/Extraction_Haskell_String_12258.out
@@ -11,6 +11,9 @@ import qualified Prelude
 
 #ifdef __GLASGOW_HASKELL__
 import qualified GHC.Base
+#if __GLASGOW_HASKELL__ >= 900
+import qualified GHC.Exts
+#endif
 #else
 -- HUGS
 import qualified IOExts
@@ -18,7 +21,11 @@ import qualified IOExts
 
 #ifdef __GLASGOW_HASKELL__
 unsafeCoerce :: a -> b
+#if __GLASGOW_HASKELL__ >= 900
+unsafeCoerce = GHC.Exts.unsafeCoerce#
+#else
 unsafeCoerce = GHC.Base.unsafeCoerce#
+#endif
 #else
 -- HUGS
 unsafeCoerce :: a -> b

